File: src/py_client_service/py_client_service/add_three_ints_service.py
from tutorial_interfaces.srv import AddThreeInts

import rclpy
from rclpy.node import Node


class MinimalService(Node):

    def __init__(self):
        # creates node with name 'minimal_service'
        super().__init__('minimal_service')
        # creates service, defining name, type and callback
        self.srv = self.create_service(AddThreeInts, 'add_three_ints', self.add_three_ints_callback)

    def add_three_ints_callback(self, request, response):
        # take the integers from the request to sum and store the result in the response
        response.sum = request.a + request.b + request.c
        self.get_logger().info('Incoming request\na: %d b: %d c: %d' % (request.a, request.b, request.c))

        return response


def main():
    # initialize ROS python client library
    rclpy.init()

    minimal_service = MinimalService()

    rclpy.spin(minimal_service)

    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/cpp_client_service/src/add_three_ints_service.cc
#include "rclcpp/rclcpp.hpp"
#include "tutorial_interfaces/srv/add_three_ints.hpp"

#include <memory>

void add(const std::shared_ptr<tutorial_interfaces::srv::AddThreeInts::Request> request, std::shared_ptr<tutorial_interfaces::srv::AddThreeInts::Response> response) {
  // add three integers from the request and store result in repsonse
  response->sum = request->a + request->b + request-> c;
  // log to console
  RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "Incoming request\na: %ld" " b: %ld" " c: %ld", request->a, request->b, request->c);
  RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "sending back response: [%ld]", (long int)response->sum);
}

int main(int argc, char **argv) {
  // initialize ROS2 c++ client library
  rclcpp::init(argc, argv);
  // create node with the name "add_two_ints_server"
  std::shared_ptr<rclcpp::Node> node = rclcpp::Node::make_shared("add_three_ints_server");

  // create service and advertise it over network
  rclcpp::Service<tutorial_interfaces::srv::AddThreeInts>::SharedPtr service = node->create_service<tutorial_interfaces::srv::AddThreeInts>("add_three_ints", &add);

  RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "Ready to add three ints.");

  // start the node, making service available for use
  rclcpp::spin(node);
  rclcpp::shutdown();
}
